flow/display_list_utils.cc

```diff
@@ -397,21 +396,71 @@ DisplayListBoundsCalculator::SaveLayerWithPaintInfo::SaveLayerWithPaintInfo(
     DisplayListBoundsCalculator* calculator,
     BoundsAccumulator* accumulator,
     const SkMatrix& saveMatrix,
+    const SkRect* saveBounds,
     const SkPaint& savePaint)
     : SaveLayerInfo(accumulator, saveMatrix),
       calculator_(calculator),
+      bounds_(saveBounds
+                  ? (saveBounds->isEmpty() ? SkRect::MakeEmpty() : *saveBounds)
+                  : kMissingBounds),
       paint_(savePaint) {}
 
+static bool PaintPreservesTransparency(const SkPaint& paint) {
+  SkImageFilter* image_filter = paint.getImageFilter();
+  // SkImageFilter::canComputeFastBounds tests for transparency behavior
+  if (image_filter && !image_filter->canComputeFastBounds()) {
+    return false;
+  }
+  SkColorFilter* color_filter = paint.getColorFilter();
+  if (color_filter &&
+      color_filter->filterColor(SK_ColorTRANSPARENT) != SK_ColorTRANSPARENT) {
+    return false;
+  }
+
+  const auto blend_mode = paint.asBlendMode();
+  if (!blend_mode) {
+    return false;  // can we query other blenders for this?
+  }
+  // Unusual blendmodes require us to process a saved layer
+  // even with operations outisde the clip.
+  // For example, DstIn is used by masking layers.
+  // https://code.google.com/p/skia/issues/detail?id=1291
+  // https://crbug.com/401593
+  switch (blend_mode.value()) {
+    // For each of the following transfer modes, if the source
+    // alpha is zero (our transparent black), the resulting
+    // blended alpha is not necessarily equal to the original
+    // destination alpha.
+    case SkBlendMode::kClear:
+    case SkBlendMode::kSrc:
+    case SkBlendMode::kSrcIn:
+    case SkBlendMode::kDstIn:
+    case SkBlendMode::kSrcOut:
+    case SkBlendMode::kDstATop:
+    case SkBlendMode::kModulate:
+      return false;
```

Comment:
A minor suggestion here would be default to more the conservative answer when encountering unknown blend modes. So list the modes that you know preserve transparency explicitly and return `true` there and default to `false` everywhere else.

Reply:
As I mentioned above, this is copied from a Skia method used for the same purpose in computing bounds for saveLayers in SkPictures. I suppose that is better, but I was hoping to delete this method when they expose this in their own code.

Reply:
Actually, if we later move away from using an SkPaint for bounds calculations here, we'd need our own method anyway, so better to keep it as defensive as possible.

Reply:
I now explicitly list all values and removed both the default case and the return statement. Now the compiler will generate an error if a value is missing. (some IDEs may show it as red as well, but the important thing is that our compiler won't accept it unless all cases are listed).